src/graph/optimizer/rule/RemoveNoopProjectRule.cpp

```diff
@@ -24,41 +71,25 @@ RemoveNoopProjectRule::RemoveNoopProjectRule() {
 }
 
 const Pattern& RemoveNoopProjectRule::pattern() const {
-  static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kProject);
+  static Pattern pattern =
+      Pattern::create(graph::PlanNode::Kind::kProject, {Pattern::create(kQueries)});
```

Comment:
What's the reason to replace the old pattern with this enumeration patterns?

Reply:
Now we will collect boundary of matched result and do some checking based on it, so we can't modify the plan nodes outside matched result.

Reply:
I think u should relax the matched result validation rather than altering the patterns. If implemented like the PR, this will be very hard to maintain many rules since people will not update these related rules when they update plan nodes in the future.

Reply:
It's a shortage, but also contains in previous implementation. It will check kind of the unmatched plan nodes too.

Reply:
And modifying unmatched nodes is anti-intuition

Reply:
Why to update the unmatched nodes? The thing this rule only needs to do is lifting the depended group node to the group containing the project node. And the depended group node needs to be cloned. So this rule don't changes the original plan.

Reply:
I means it accesses, clones and replaces the unmatched nodes (dependencies of Project).

Reply:
so what?

Reply:
It's not good practice.